server/aesdsocket.c

```diff
@@ -0,0 +1,377 @@
+        /* It redirects the standard input, output, and error to /dev/null by opening and duplicating file descriptors, 
+         * effectively detaching the process from the terminal and eliminating any input or output connections with the daemon.
+         */
+        /* redirect fd's 0,1,2 to /dev/null */
+		open ("/dev/null", O_RDWR);     /* stdin */
+        dup (0);                        /* stdout */
+        dup (0);                        /* stderror */
+	}    
+
+    //Create file to push data to /var/tmp/aesdsocketdata
+    data_file_fd = open(DATA_FILE,O_CREAT|O_RDWR,0777);
+	if(-1 == data_file_fd)
+	{
+	    perror("Error: Couldn't open the file at /var/temp/aesdsocketdata");
+		exit(ERROR_CODE);
+	}
+
+    if(-1 == listen(main_sockfd,BACKLOG)) //backlog assumed 10 i.e, >1
+    {
+        perror("listen()");
+        freeaddrinfo(res);
+        close(main_sockfd);
+        exit(ERROR_CODE);
+    }
+    freeaddrinfo(res); // all done with this structure
+    /*
+     *    Need a while(1) so that we can accept many client requests and handle them. 
+     */
+    while(1)
+    {
+        socklen_t client_len = sizeof(client_addr);
+        client_sockfd = accept(main_sockfd,(struct sockaddr*)&client_addr,&client_len);
+
+        if (-1 == client_sockfd) 
+        {
+            perror("accept()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+
+        /*
+         *  Reference for  inet_ntop : https://man7.org/linux/man-pages/man3/inet_ntop.3.html
+         */ 
+        if(NULL == inet_ntop(AF_INET, get_in_addr((struct sockaddr*)&client_addr),ip_addr, sizeof(ip_addr)))
+        {
+            perror("inet_ntop()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+        else
+        {
+            syslog(LOG_DEBUG,"Accepted connection from %s", ip_addr);
+        }
+
+        char *data_buf = (char*)calloc(initial_alloc_size,sizeof(char));
+        if(NULL == data_buf)
+        {
+            perror("calloc()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+
+        /*
+         * For completing any open connection operations i.e., resv() and send(), need to block the signal using sigprocmask()
+         * Blocking doesn't mean we miss the signal, the signal waits until its unblocked and gets serviced by kernel. 
+         */
+		if( -1 == sigprocmask(SIG_BLOCK, &sock_set, NULL))
+        {
+			perror("sigprocmask()");
+			error_handler();
+			exit(ERROR_CODE);
+        }
+
+        while((bytes_read = recv(client_sockfd,data_buf+present_location,initial_alloc_size,0)) > 0)
+        {
+
+            if(-1 == bytes_read)
+            {
+                perror("resv()");
+                error_handler();
+                exit(ERROR_CODE);          
+            }            
+            /*
+             * Reference to strchr function : https://man7.org/linux/man-pages/man3/strchr.3.html
+            */
+            if(strchr(data_buf ,'\n') != NULL)
```

Comment:
@Rishikesh38 `calloc` handles null initialization for you but what would happen if you received exactly `initial_alloc_size` bytes here?  How would you ensure null termination in this case?

Reply:
It would be better if I take care of null character than leaving it to the alloc functions.